File: backend/app/services/package_status.py
"""
Package status transition validation and service.
Enforces strict status progression: MATCHED → PICKED_UP → IN_TRANSIT → DELIVERED
"""
from datetime import datetime
from typing import Tuple, List, Optional
from sqlalchemy.orm import Session

from app.models.package import Package, PackageStatus
from app.models.user import User


# Define allowed status transitions
ALLOWED_TRANSITIONS = {
    PackageStatus.PENDING: [PackageStatus.MATCHED, PackageStatus.CANCELLED],
    PackageStatus.MATCHED: [PackageStatus.PICKED_UP, PackageStatus.PENDING, PackageStatus.CANCELLED],
    PackageStatus.PICKED_UP: [PackageStatus.IN_TRANSIT],
    PackageStatus.IN_TRANSIT: [PackageStatus.DELIVERED],
    PackageStatus.DELIVERED: [],  # Terminal state
    PackageStatus.CANCELLED: [],  # Terminal state
}

# Human-readable status names for error messages
STATUS_LABELS = {
    PackageStatus.PENDING: "Pending",
    PackageStatus.MATCHED: "Matched",
    PackageStatus.PICKED_UP: "Picked Up",
    PackageStatus.IN_TRANSIT: "In Transit",
    PackageStatus.DELIVERED: "Delivered",
    PackageStatus.CANCELLED: "Cancelled",
}

# Required progression path for delivery
DELIVERY_PATH = [
    PackageStatus.PENDING,
    PackageStatus.MATCHED,
    PackageStatus.PICKED_UP,
    PackageStatus.IN_TRANSIT,
    PackageStatus.DELIVERED,
]


def validate_transition(current: PackageStatus, target: PackageStatus) -> Tuple[bool, str]:
    """
    Validate if a status transition is allowed.

    Args:
        current: Current package status
        target: Target package status

    Returns:
        Tuple of (is_valid, error_message)
    """
    if current == target:
        return False, f"Package is already in {STATUS_LABELS[current]} status"

    if target not in ALLOWED_TRANSITIONS.get(current, []):
        allowed = ALLOWED_TRANSITIONS.get(current, [])
        if not allowed:
            return False, f"Cannot change status from {STATUS_LABELS[current]} - this is a terminal state"

        allowed_names = [STATUS_LABELS[s] for s in allowed]
        return False, (
            f"Invalid status transition: Cannot go from {STATUS_LABELS[current]} to {STATUS_LABELS[target]}. "
            f"Allowed next statuses: {', '.join(allowed_names)}"
        )

    return True, ""


def get_allowed_next_statuses(current: PackageStatus) -> List[str]:
    """
    Get list of allowed next statuses for a given current status.

    Args:
        current: Current package status

    Returns:
        List of allowed status values
    """
    return [s.value for s in ALLOWED_TRANSITIONS.get(current, [])]


def can_mark_delivered(package: Package) -> Tuple[bool, str]:
    """
    Check if a package can be marked as delivered.
    Verifies proof requirement and current status.

    Args:
        package: The package to check

    Returns:
        Tuple of (can_deliver, error_message)
    """
    # Must be in IN_TRANSIT status
    if package.status != PackageStatus.IN_TRANSIT:
        return False, (
            f"Package must be In Transit before marking as Delivered. "
            f"Current status: {STATUS_LABELS[package.status]}"
        )

    # If proof is required, check that delivery proof exists
    if package.requires_proof:
        # This will be checked by the delivery proof service
        # Here we just indicate that proof is required
        return True, "proof_required"

    return True, ""


def transition_package(
    db: Session,
    package: Package,
    target_status: PackageStatus,
    actor_id: int,
    force: bool = False
) -> Tuple[Package, str]:
    """
    Transition a package to a new status with validation.

    Args:
        db: Database session
        package: The package to transition
        target_status: The target status
        actor_id: ID of the user performing the action
        force: If True, skip validation (admin only)

    Returns:
        Tuple of (updated_package, error_message)
        If error_message is not empty, the transition failed
    """
    if not force:
        is_valid, error = validate_transition(package.status, target_status)
        if not is_valid:
            return package, error

    # Record the transition
    now = datetime.utcnow()
    old_status = package.status
    package.status = target_status
    package.status_changed_at = now

    # Update specific timestamp fields based on target status
    if target_status == PackageStatus.MATCHED:
        package.matched_at = now
        package.courier_id = actor_id  # Set courier when matched
    elif target_status == PackageStatus.PICKED_UP:
        package.picked_up_at = now
        package.pickup_time = now  # Also update legacy field
    elif target_status == PackageStatus.IN_TRANSIT:
        package.in_transit_at = now
    elif target_status == PackageStatus.DELIVERED:
        package.delivery_time = now
    elif target_status == PackageStatus.PENDING:
        # Courier declined - clear courier assignment
        if old_status == PackageStatus.MATCHED:
            package.courier_id = None
            package.matched_at = None

    db.commit()
    db.refresh(package)

    return package, ""


def get_status_progress(package: Package) -> dict:
    """
    Get the progress of a package through the delivery lifecycle.

    Args:
        package: The package to check

    Returns:
        Dict with progress information
    """
    current_index = -1
    for i, status in enumerate(DELIVERY_PATH):
        if package.status == status:
            current_index = i
            break

    # Handle cancelled separately
    if package.status == PackageStatus.CANCELLED:
        return {
            "current_step": -1,
            "total_steps": len(DELIVERY_PATH) - 1,  # Exclude PENDING from count
            "is_terminal": True,
            "is_cancelled": True,
            "progress_percent": 0,
            "steps": _build_steps(package, -1),
        }

    return {
        "current_step": current_index,
        "total_steps": len(DELIVERY_PATH) - 1,  # Exclude PENDING from count
        "is_terminal": package.status in [PackageStatus.DELIVERED, PackageStatus.CANCELLED],
        "is_cancelled": False,
        "progress_percent": int((current_index / (len(DELIVERY_PATH) - 1)) * 100) if current_index > 0 else 0,
        "steps": _build_steps(package, current_index),
    }


def _build_steps(package: Package, current_index: int) -> List[dict]:
    """Build step information for progress display."""
    steps = []
    timestamps = {
        PackageStatus.PENDING: package.created_at,
        PackageStatus.MATCHED: package.matched_at,
        PackageStatus.PICKED_UP: package.picked_up_at,
        PackageStatus.IN_TRANSIT: package.in_transit_at,
        PackageStatus.DELIVERED: package.delivery_time,
    }

    for i, status in enumerate(DELIVERY_PATH):
        if status == PackageStatus.PENDING:
            continue  # Skip pending in display

        timestamp = timestamps.get(status)
        steps.append({
            "status": status.value,
            "label": STATUS_LABELS[status],
            "completed": i <= current_index if current_index >= 0 else False,
            "current": i == current_index,
            "timestamp": timestamp.isoformat() if timestamp else None,
        })

    return steps

File: packages/shared-utils/src/package-status.ts
// Package status state machine
// Derived from backend/app/services/package_status.py

export type PackageStatus =
  | 'NEW'
  | 'OPEN_FOR_BIDS'
  | 'BID_SELECTED'
  | 'PENDING_PICKUP'
  | 'IN_TRANSIT'
  | 'DELIVERED'
  | 'CANCELED'
  | 'FAILED'

export const ALLOWED_TRANSITIONS: Record<PackageStatus, PackageStatus[]> = {
  NEW: ['OPEN_FOR_BIDS', 'CANCELED'],
  OPEN_FOR_BIDS: ['BID_SELECTED', 'CANCELED'],
  BID_SELECTED: ['PENDING_PICKUP', 'OPEN_FOR_BIDS', 'CANCELED'],
  PENDING_PICKUP: ['IN_TRANSIT', 'CANCELED'],
  IN_TRANSIT: ['DELIVERED', 'FAILED'],
  DELIVERED: [],
  CANCELED: [],
  FAILED: ['OPEN_FOR_BIDS'], // Admin only
}

export const STATUS_LABELS: Record<PackageStatus, string> = {
  NEW: 'New',
  OPEN_FOR_BIDS: 'Open for Bids',
  BID_SELECTED: 'Bid Selected',
  PENDING_PICKUP: 'Pending Pickup',
  IN_TRANSIT: 'In Transit',
  DELIVERED: 'Delivered',
  CANCELED: 'Canceled',
  FAILED: 'Failed',
}

export const STATUS_COLORS: Record<PackageStatus, string> = {
  NEW: 'gray',
  OPEN_FOR_BIDS: 'blue',
  BID_SELECTED: 'purple',
  PENDING_PICKUP: 'yellow',
  IN_TRANSIT: 'orange',
  DELIVERED: 'green',
  CANCELED: 'red',
  FAILED: 'red',
}

export const TERMINAL_STATES: PackageStatus[] = ['DELIVERED', 'CANCELED']

/**
 * Check if a status is a terminal state (no further transitions possible)
 */
export function isTerminalState(status: PackageStatus): boolean {
  return TERMINAL_STATES.includes(status)
}

/**
 * Check if a package can be canceled from its current status
 */
export function canCancel(status: PackageStatus, isAdmin: boolean = false): boolean {
  if (TERMINAL_STATES.includes(status)) return false
  if (isAdmin) return true
  if (status === 'FAILED' || status === 'IN_TRANSIT') return false
  return true
}

/**
 * Get the allowed next statuses for a package
 */
export function getAllowedNextStatuses(current: PackageStatus, isAdmin: boolean = false): PackageStatus[] {
  const allowed = ALLOWED_TRANSITIONS[current] || []
  if (current === 'FAILED' && !isAdmin) return []
  if (!isAdmin) {
    return allowed.filter((s) => s !== 'FAILED')
  }
  return allowed
}

/**
 * Check if a transition from one status to another is valid
 */
export function isValidTransition(from: PackageStatus, to: PackageStatus, isAdmin: boolean = false): boolean {
  const allowed = getAllowedNextStatuses(from, isAdmin)
  return allowed.includes(to)
}

/**
 * Get the human-readable label for a status
 */
export function getStatusLabel(status: string): string {
  return STATUS_LABELS[status.toUpperCase() as PackageStatus] || status
}

/**
 * Get the color associated with a status (for UI badges)
 */
export function getStatusColor(status: string): string {
  return STATUS_COLORS[status.toUpperCase() as PackageStatus] || 'gray'
}
